[#8] 세션 로그인 , 인증 및 인가

sukeun1997 opened the pull request:
## 커밋 내용

- 사용자 , 사장 , 라이더 테이블 통합 = User
- 로그인 , 로그아웃 구현 
- 커스텀 애노테이션과 AOP 를 사용하여 인증 및 인가 구현
- 테스트
- 피드백 반영

## 작업 예정

- 피드백 기반 코드 리팩토링 진행
- 세션 클러스터링 진행
- 유저 부분 추가 기능 구현 ( 정보변경, 회원 탈퇴 ..)

sukeun1997 committed:
- schema.sql 추가
- MyBatis 설정 추가

sukeun1997 committed:
- 유저 회원가입 기반 추가
- 의존성 추가

sukeun1997 committed:
Test :  UserMapper 기능 추가 및 테스트

sukeun1997 committed:
Test : 회원가입 테스트

sukeun1997 committed:
Feat : 회원가입 및 예외처리

Test : 회원가입

sukeun1997 committed:
Feat : 회원가입 템플릿콜백 적용 및 사장님 회원가입

Test : 사장님 회원가입

sukeun1997 committed:
Refactor : 테스트 공통부분 추상화

sukeun1997 committed:
Feat : 라이더 회원가입

sukeun1997 committed:
Feat : 패스워드 암호화 적용

sukeun1997 committed:
Refactor : Controller 응답 부분 중복 제거 및 Dto 정리

sukeun1997 committed:
Feat : 아이디 중복체크

sukeun1997 committed:
Refactor : 메서드 명 및 클래스명 수정

sukeun1997 committed:
Feat : 유저 로그인

Test : 유저 로그인

sukeun1997 committed:
Feat : 사장 , 라이더 로그인

sukeun1997 committed:
Feat : 회원 로그아웃

sukeun1997 committed:
Clean : Erd 에 맞춰 구조 변경

sukeun1997 committed:
Fix : 데이터베이스 테이블 수정 및 코드 반영

Refactor : 회원 level 테이블 추가

sukeun1997 committed:
Feat : 로그인 여부 확인

sukeun1997 committed:
Feat : 사용자 인증 및 인가 확인

sukeun1997 committed:
Refactor : 인증, 인가 에 따른 오류 코드 변경

sukeun1997 committed:
Rename : Certify -> Auth

Refactor : 쿼리문 소문자 -> 대문자로 변경

f-lab-jason commented on .gitignore:
```diff
@@ -25,6 +25,7 @@ bin/
 out/
 !**/src/main/**/out/
 !**/src/test/**/out/
+client/**
```
요건 어떤 용도인가요?

sukeun1997 replied:
`.http` 로 API 테스트 할 때 사용해서 gitignore 로 제외 시켰습니다

f-lab-jason commented on src/main/java/com/flab/delivery/annotation/hasAuthorization.java:
```diff
@@ -0,0 +1,17 @@
+    UserLevel level();
+
+    enum UserLevel {
+        ALL, USER, RIDER, OWNER
```
(nit) enum은 한 라인에 하나씩 적어주시면 좋겠습니다 :)

f-lab-jason commented on src/main/java/com/flab/delivery/annotation/hasAuthorization.java:
```diff
@@ -0,0 +1,17 @@
+
+    UserLevel level();
+
+    enum UserLevel {
```
User**Level**은 위아래가 있는 개념입니다. 예를 들어 일반사용자 < 관리자 < 전체 관리자 같은..
여기는 유저별로 기능이 다르니 UserType같은게 더 괜찮은 이름같아요. User 객체에서 type같은게 더 나을 것 같구요.

f-lab-jason commented on src/main/java/com/flab/delivery/config/AppConfig.java:
```diff
@@ -0,0 +1,15 @@
+@Configuration
+public class AppConfig {
+
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        return new JbCrypt();
+    }
```
`PasswordEncoder`자체가 유틸성 클래스이니 내부적으로 static method들만 가지는 클래스로 만들고 instantiation은 없애는게 더 나을 것 같아요.

f-lab-jason commented on src/main/java/com/flab/delivery/controller/UserController.java:
```diff
@@ -0,0 +1,57 @@
+        return STATUS_OK;
+    }
+
+    @hasAuthorization(level = ALL)
```
`level = ALL`을 생략할 수 있는 형태로 하는건 어떨까요?
level을 명시하지 않으면 ALL과 같은 동작을 하도록

f-lab-jason commented on src/main/java/com/flab/delivery/dto/LoginDto.java:
```diff
@@ -0,0 +1,25 @@
+    @NotBlank(message = "비밀번호를 입력해 주세요.")
+    private String password;
+
+    public LoginDto(String id, String password) {
+        this.id = id;
+        this.password = password;
+    }
```
`@AllArgsConstructor`대신 따로 적은 이유가 있을까요?

sukeun1997 replied:
테스트에서 필요하여 필드 변수 개수가 적어서 생성자를 사용했습니다. 필드 변수 개수가 적어도 `@AllargsConstructor` 를 사용하는게 더 나을까요 ??  추가적으로 `@AllargsConstructor`  를 사용하면 현재 `id` 와 `password` 에 필드 순서가 바뀌더라도 컴파일 시점에서 확인할 수 있는 방법이 없어 문제가 생길 수 있다고 생각하는데 현재 LoginDto 에는 `@NoArgsConstructor(access = AccessLevel.PRIVATE)` 이 있어  `@Builder` 사용이 불가능 하므로 `package-private` 생성자를 만들고 `@Builder` 를 사용 하는게 맞다고 생각이 드는데 제 생각이 맞을까요 !?

f-lab-jason commented on src/main/java/com/flab/delivery/dto/UserDto.java:
```diff
@@ -0,0 +1,36 @@
+    @NonNull
+    private String level;
+
+    @Builder
+    public UserDto(@NonNull String id, @NonNull String password, @NonNull String name, @NonNull String email, @NonNull String phoneNumber, @NonNull String level) {
+        this.id = id;
+        this.password = password;
+        this.name = name;
+        this.email = email;
+        this.phoneNumber = phoneNumber;
+        this.level = level;
+    }
```
클래스 레벨에 `@Builder`를 선언한것과 어떻게 다른가요?

sukeun1997 replied:
클래스 레벨에 선언하면 컴파일 시점에 모든 필드에 대한 `package-private` 생성자를 생성하여 빌더를 생성하고, 생성자에 붙이면 `생성자에 있는 매개변수`에 맞춰 빌더가 생성됩니다.

f-lab-jason commented on src/main/java/com/flab/delivery/mapper/UserMapper.java:
```diff
@@ -0,0 +1,19 @@
+
+    void save(SignUpDto signUpDto);
+
+    boolean existsUserById(String id);
```
(nit) 개인적으로 `hasUserById`를 더 선호합니다 :)

f-lab-jason commented on src/main/java/com/flab/delivery/service/SessionLoginService.java:
```diff
@@ -0,0 +1,31 @@
+@RequiredArgsConstructor
+public class SessionLoginService implements LoginService {
+
+    private static final String SESSION_ID = "SESSION_ID";
```
이곳 외에 테스트에서도 쓰이니 Constants같은 클래스를 하나 만들어서 상수들만 따로 보관해주면 좋겠어요.

f-lab-jason commented on src/main/java/com/flab/delivery/service/UserService.java:
```diff
@@ -0,0 +1,57 @@
+        mapper.save(signUpDto);
+    }
+
+    public void checkIdDuplicated(String id) {
```
`checkIdDuplication` 혹은 `checkDuplicatedId`이 더 나아보입니다.

f-lab-jason commented on src/main/java/com/flab/delivery/service/UserService.java:
```diff
@@ -0,0 +1,57 @@
+    public void checkIdDuplicated(String id) {
+        if (mapper.existsUserById(id)) {
+            log.error("유저 회원가입 이미 존재하는 아이디 =  {} ", id);
+            throw new UserException("이미 존재하는 아이디 입니다.");
```
(nit) `BadInputException`에 더 가깝지 않을까요?

f-lab-jason commented on src/main/java/com/flab/delivery/utils/PasswordEncoder.java:
```diff
@@ -0,0 +1,8 @@
+
+public interface PasswordEncoder {
+
+    String encoder(String password);
```
`encode(...)`

f-lab-jason commented on src/main/java/com/flab/delivery/utils/PasswordEncoder.java:
```diff
@@ -0,0 +1,8 @@
+
+    String encoder(String password);
+
+    boolean isMatch(String password, String hashValue);
```
`matches(...)`

f-lab-jason commented on src/main/resources/mapper/UserMapper.xml:
```diff
@@ -0,0 +1,25 @@
+    </insert>
+
+    <select id="existsUserById" resultType="boolean">
+        SELECT exists(SELECT id FROM user WHERE id = #{id});
```
(nit) 현재같은 유저베이스는 크게 중요하진 않지만 중복체크할땐 TOP 1을 가져오는게 제일 퍼포먼스는 좋습니다. (요런 케이스는 어차피 인덱스 탈꺼라 크게 차이안나긴해요.)

f-lab-jason commented on src/main/resources/mapper/UserMapper.xml:
```diff
@@ -0,0 +1,25 @@
+        SELECT exists(SELECT id FROM user WHERE id = #{id});
+    </select>
+
+    <select id="getCountById" resultType="long">
+        SELECT count(id)
+        FROM user;
+    </select>
```
이건 무슨 용도인가요?

sukeun1997 replied:
현재는 테스트에서 Insert , Delete 에 의한 회원 수 체크에서 사용하고 있습니다

f-lab-jason reviewed (commented):
1. 왕규님과 상의하셔서 code formatter를 적용하면 좋겠습니다. (구글이 제일 만만) double empty lines 같은걸 잡는데 도움될거예요!
2. ResponseEntity는 굳이 static으로 만들어놓을 필요 없이 ResponseEntity.ok()같은걸로 생성하셔도 무방합니다.
3. 왕규님과 상의해서 API 응답 포맷을 만들면 좋겠어요. 보통 API version, status, message, data(리턴되는 오브젝트) 등을 포함하는 객체를 하나 만들어서 씁니다.

sukeun1997 committed:
Refactor : 피드백 반영

sukeun1997 committed:
Refactor : 피드백 반영

Remove : HttpStatusResponse.java
Add : 입력값에 의한 예외 추가

sukeun1997 committed:
Reafactor : 피드백 반영

Add : 공통 응답 포맷 추가 및 적용
Fix : @AllArgsConsturcotr accessLevel PUBLIC -> PRIVATE 로 변경

cwangg897 commented on src/test/java/com/flab/delivery/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,189 @@
+        // then
+        mockMvc.perform(get(getExistsUri(signUpDto)))
+                .andExpect(jsonPath("$.status").value(400))
+                .andExpect(jsonPath("$.message").exists())
```
😀안녕하세요, 수근님 
UserService에서 예외메시지를 보내면서 예외를 보내고 계시는데, 메시지 존재만 확인보다 저의 생각은 메시지 내용도 확인하는게 좋다고생각합니다  저의 의견에 대한 수근님 의견이 궁금합니다!!!

cwangg897 commented on src/main/java/com/flab/delivery/controller/UserController.java:
```diff
@@ -0,0 +1,53 @@
+    private final UserService userService;
+
+    @PostMapping
+    public CommonResult signUp(@RequestBody @Valid SignUpDto signUpDto) {
```
수근님 저의 의견이있습니다
Raw타입보다 타입 파라미터를 명시해주는게 좋다고생각합니다 
왜냐하면 Raw 타입을 사용하면 컴파일 시점에 문제를 잡지 못하다가 런타임 시점에 ClassCastException 에러가 발생할 수 있습니다
하지만 만약 Raw 타입이 아니라 타입 파라미터를 명시해준다면, 컴파일 시점에 오류를 잡아낼 수 있습니다.
그렇기 때문에 위와 같은 Raw 타입의 반환을 피하고 반환할 데이터 타입을 명시하는 것이 좋다고 생각합니다
수근님 의견은 어떠신가요?

f-lab-jason reviewed (approved):
🚢 
왕규님 리뷰 내용에 대해서만 확인해주시면 좋겠습니다!
